src/backend/utils_internal_gpu/cuMovemem_gpu.cu

```diff
@@ -23,11 +23,21 @@ namespace cytnx {
 #ifdef UNI_GPU
     template <class BidirectionalIterator>
     void reverse_perm(BidirectionalIterator first, BidirectionalIterator last, int N) {
-      while ((first != last) && (first != --last)) {
+      // while ((first != last) && (first != --last)) {
+      //   *first = (N - 1) - *first;
+      //   *last = (N - 1) - *last;
+      //   std::iter_swap(first, last);
+      //   ++first;
+      // }
+      // if (N % 2) *first = (N - 1) - *first;
+      --last;
+      while (first != last and first != --last) {
+        ++last;
         *first = (N - 1) - *first;
         *last = (N - 1) - *last;
         std::iter_swap(first, last);
         ++first;
+        --last;
```

Comment:
The original for loop is correct. See https://godbolt.org/z/bca7c35e7

Reply:
restored